### PR TITLE
Add positional wildcard support to module zos_job_query

### LINKED ISSUE #323: [Feature] [zos_job_query]More asterisk support for z/OS ansible modules
I have a playbook using zos_job_query to query jobname "CPY*RS*"

      zos_job_query:
        job_name: "CPY*RS*"

but it failed and report the following error:
Value of unknown type: <class 'RuntimeError'>, Failed to validate the job name: CPY*RS*\n

      zos_job_query:
        job_name: "CPY*"

Describe the solution you'd like:
It looks like only support asterisk * as the suffix of the jobname now.
It would be better to have more than 1 asterisk support and even other wildcard support.

### PR DESCRIPTION
**Pull request by richp405**
##### SUMMARY
Adds embedded and multiple asterisk support to zos_job_query
Fixes #323 
##### ISSUE TYPE
Feature pull request AND bugfix pull request

##### COMPONENT NAME
zos_job_query showed the problem.  Actual issue was is search condition of job/_get_job_status

##### ADDITIONAL INFORMATION
If the job id or name is more than an asterisk (example look for "JOB003*), or multiple asterisk (J*003*)
This caused limited results and potential failures, because the asterisks didn't work correctly
This has been changed to use fnmatch (note: this IS case insensitive)

examples:
 - name: Check for I*I* embedded star
      zos_job_query:
        job_name: I*I*
      register: result

 - name: Check for STC*3* job embedded star
      zos_job_query:
        job_id: STC*3*
      register: result

```paste below

```

### CODE REVIEW

File: plugins/modules/zos_job_query.py
Comment (fernandofloresg) [inline plugins/modules/zos_job_query.py:227]: job_name_final appears to not be used after this assignment

File: plugins/modules/zos_job_query.py
Comment (fernandofloresg) [inline plugins/modules/zos_job_query.py:1]: variable names o, m, n make difficult for maintainers to read over the code, I had to read over it multiple times to get the idea. Could you give them more descriptive names?

File: tests/functional/modules/test_zos_job_query_func.py
Comment (fernandofloresg) [inline tests/functional/modules/test_zos_job_query_func.py:1]: This test file seems missing cases, is there a way to add a positive test case? Running a couple of jobs and looking for them?

**Review by fernandofloresg — CHANGES_REQUESTED** [review]
Overall change looks good to me, comments were added in the review

File: changelogs/fragments/323-zos-job-query-handle-multiple-wildcards.yml
Comment (richp405) [reply changelogs/fragments/323-zos-job-query-handle-multiple-wildcards.yml:1]: Added.

File: plugins/modules/zos_job_query.py
Comment (richp405) [reply plugins/modules/zos_job_query.py:227]: forgot to remove it... was part of a test set.

File: plugins/modules/zos_job_query.py
Comment (richp405) [reply plugins/modules/zos_job_query.py:279]: yep

File: plugins/modules/zos_job_query.py
Comment (richp405) [reply plugins/modules/zos_job_query.py:1]: Changed to test_basic, test_star and test_result.  Hope that clears it up... Sorry for the 'letters'... it's a bad habit from my ancient c days.  Thanks for pointing it out.

File: tests/functional/modules/test_zos_job_query_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_job_query_func.py:1]: The original case in there, which was 'list all jobs' was the positive case... I added an addition test of wildcards agains job_id in addition to job_name.  Would that work, or do you think I should run and test a specific job (since that's done in other tests)?

File: plugins/modules/zos_job_query.py
Comment (fernandofloresg) [reply plugins/modules/zos_job_query.py:1]: I'd imagine. Thanks for changing it but comments still have the letters

File: tests/functional/modules/test_zos_job_query_func.py
Comment (fernandofloresg) [reply tests/functional/modules/test_zos_job_query_func.py:1]: I would say yes to make sure we find them. Btw ran the Jenkins pipeline and test "test_zos_job_query_multi_wildcards_func" failed :

Screenshot 2023-04-03 at 3 24 23 PM

File: plugins/modules/zos_job_query.py
Comment (richp405) [reply plugins/modules/zos_job_query.py:1]: ack!!! I fixed the code and skipped the comments.

File: tests/functional/modules/test_zos_job_query_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_job_query_func.py:1]: When copying the test, I didn't rename it.  I'll fix and re-run the afternoon.

File: plugins/modules/zos_job_query.py
Comment (ddimatos) [inline plugins/modules/zos_job_query.py:256]: @richp405 - is error meant for users , seems very specific to the code, eg `ix was` , users will not know what `ix` is.  I think just returning the value they gave us is enough, no need to share with them the logic of short and ix. 

Wondering if it should be "Unable to find the job name: {0}".format(str(job_name_in))

File: plugins/modules/zos_job_query.py
Comment (ddimatos) [inline plugins/modules/zos_job_query.py:245]: m/n comments don't map to the updated code , its minor just might be helpful for someone in the future.

File: tests/functional/modules/test_zos_job_query_func.py
Comment (ddimatos) [inline tests/functional/modules/test_zos_job_query_func.py:48]: I don't see  a job submitted thus we are relying on the ones on the system which works for now but not for portability in the future, but this is easy enough to fix. What caught my eye was the  `job_name="JOB*1*"` seems to be a JOB ID, i logged on to my system to find a job name starting with JOB but all I found was Job IDs. Could this be an error?

**Comment by ddimatos** [comment]
We should update our doc further highlighting the now supported wildcard placement.

**Review by ddimatos — COMMENTED** [review]
We should update our doc further highlighting the now supported wildcard placement.

File: tests/functional/modules/test_zos_job_query_func.py
Comment (ddimatos) [inline tests/functional/modules/test_zos_job_query_func.py:38]: Are we able to support job names that start with * , what would that look like if we do for very busy system?

**Review by ddimatos — CHANGES_REQUESTED** [review]
I don't know where to put these next two comments as there is no change for me to comment where I want so I will add them here.

1) We should update our doc further highlighting the now supported wildcard placement.
2) We should add some examples to the doc to highlight the capabilities now, some examples with Job Names, Job Ids wild card placement.

File: tests/functional/modules/test_zos_job_query_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_job_query_func.py:53]: Yes... not as different as you might think: regardless of the mask, we ask for something like jls *, then filter the results in the python.  If it's job_id, we get a smaller list, up to the first asterisk; it it's job_name, we get them all.

File: tests/functional/modules/test_zos_job_query_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_job_query_func.py:1]: I fixed, and expanded both cases to run a 'hello' jcl, then check for the results.  Haven't gotten through pipeline, as someone else is running it, but will try again at end of day

File: plugins/modules/zos_job_query.py
Comment (richp405) [reply plugins/modules/zos_job_query.py:256]: That's simpler and better.  Making the change.

File: tests/functional/modules/test_zos_job_query_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_job_query_func.py]: Yes..Fernando mentioned this as well... the new tests run a copy of the hello jcl.  Job ID takes out 2 characters and uses asterisks for search, while job name search looks for "HE*L*".... That should go through the pipeline this evening.

**Comment by richp405** [comment]
@ddimatos I didn't see a widget to reply to your comment about documentation, so I'm adding this at the end... I tweaked (minor changes) to the documentation at the top, with examples and the data type descriptions, to expand on the asterisk support.  I'll ping you once this passes the pipeline.

File: tests/functional/modules/test_zos_job_query_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_job_query_func.py:1]: I called a function with un-named parameters.  Fixed:

image

**Review by ddimatos — APPROVED** [review]
Looks good to me @richp405 , we need one more reviewer and we are good to merge.

**Comment by ddimatos** [comment]
The last regression for this was successful 
image

**Review by fernandofloresg — APPROVED** [review]
Looks good to me. Thanks for doing all the requested changes.

**Comment by richp405** [comment]
One more edit, and through the pipeline

image

**Comment by fernandofloresg** [comment]
Thanks @richp405

**Pull request merged**
